node/src/config.rs

```diff
@@ -100,7 +122,335 @@ fn validate_name(s: &str) -> Result<()> {
+                    block_ingestor::<graph_chain_ethereum::Chain>(logger, id, chain, &mut res)
+                        .await?
+                }
+                BlockchainKind::Near => {
+                    block_ingestor::<graph_chain_near::Chain>(logger, id, chain, &mut res).await?
+                }
+                BlockchainKind::Cosmos => {
+                    block_ingestor::<graph_chain_cosmos::Chain>(logger, id, chain, &mut res).await?
+                }
+                BlockchainKind::Substreams => {
+                    // handle substreams later
+                }
+                BlockchainKind::Starknet => {
+                    block_ingestor::<graph_chain_starknet::Chain>(logger, id, chain, &mut res)
+                        .await?
+                }
+            }
+        }
+
+        // substreams networks that also have other types of chain(rpc or firehose), will have
+        // block ingestors already running.
+        let visited: Vec<_> = res.iter().map(|b| b.network_name()).collect();
+        for ((_, id), chain) in blockchain_map
+            .0
+            .iter()
+            .filter(|((kind, id), _)| BlockchainKind::Substreams.eq(&kind) && !visited.contains(id))
+        {
+            block_ingestor::<graph_chain_substreams::Chain>(logger, id, chain, &mut res).await?
+        }
+
+        Ok(res)
+    }
+
+    pub async fn blockchain_map(
+        &self,
+        config: &Arc<EnvVars>,
+        node_id: &NodeId,
+        logger: &Logger,
+        store: &Store,
+        logger_factory: &LoggerFactory,
+        metrics_registry: Arc<MetricsRegistry>,
+        chain_head_update_listener: Arc<ChainHeadUpdateListener>,
+    ) -> BlockchainMap {
+        let mut bm = BlockchainMap::new();
+
+        networks_as_chains(
+            config,
+            &mut bm,
+            node_id,
+            logger,
+            self,
+            store,
+            logger_factory,
+            metrics_registry,
+            chain_head_update_listener,
+        )
+        .await;
+
+        bm
+    }
+
+    pub fn firehose_endpoints(&self, chain_id: ChainId) -> FirehoseEndpoints {
+        FirehoseEndpoints::new(chain_id, self.firehose_provider_manager.cheap_clone())
+    }
+
+    pub fn substreams_endpoints(&self, chain_id: ChainId) -> FirehoseEndpoints {
+        FirehoseEndpoints::new(chain_id, self.substreams_provider_manager.cheap_clone())
+    }
+
+    pub fn ethereum_rpcs(&self, chain_id: ChainId) -> EthereumNetworkAdapters {
+        let eth_adapters = self
+            .adapters
+            .iter()
+            .filter(|a| a.chain_id().eq(&chain_id))
+            .flat_map(|a| a.as_rpc())
+            .flat_map(|eth_c| eth_c.call_only.clone())
+            .collect_vec();
+
+        EthereumNetworkAdapters::new(
+            chain_id,
+            self.rpc_provider_manager.cheap_clone(),
+            eth_adapters,
+            None,
+        )
+    }
+}
+
 impl Config {
+    pub async fn networks(
```

Comment:
This module was really only meant for handling the configuration file and any checks that can be performed on the config file without any other setup. It feels like some of the headaches this needs to address is that, because `Config` has very little setup around it, the code needs to work both for when nothing has been set up and for when we actually have things like a store available.

It might be better to restructure this so that all this network setup code gets run from, say, a `NetworkSetup` struct that gets a config and whatever else is needed (like the `BlockStore`) and can then do its thing knowing that the store exists already.

Reply:
This is basically what Networks is, I could move it to it's own module and add a from_config instead of having config.networks(), this was just for convenience, I don't feel strongly either way.

Reply:
Yes, I think putting it into its own module would be better. It also seems to me that requiring having a `BlockStore` here would simplify the logic.